kitchen/src/kitchen/demos.clj

```diff
@@ -0,0 +1,144 @@
+;;
+;;       Copyright (c) 2017 Two Sigma Investments, LP.
+;;       All Rights Reserved
+;;
+;;       THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF
+;;       Two Sigma Investments, LP.
+;;
+;;       The copyright notice above does not evidence any
+;;       actual or intended publication of such source code.
+;;
+(ns kitchen.demos
+  (:require [clojure.data.json :as json]
+            [clojure.java.io :as io]
+            [clojure.java.shell :as sh]
+            [clojure.string :as str]
+            [clojure.tools.logging :as log]
+            [kitchen.utils :as utils])
+  (:import (java.io File)
+           (java.net URLEncoder)))
+
+(defn- request->cid [request]
+  (get-in request [:headers "x-cid"]))
+
+(defn- request->cid-string [request]
+  (str "[CID=" (request->cid request) "]"))
+
+(defn- printlog
+  ([request & messages]
+   (log/info (request->cid-string request) (str/join " " messages))))
+
+(defn- search->image-titles
+  [search-query]
+  (let [extract-titles (fn extract-titles [page-entries]
+                         (mapcat (fn [page-entry] (map #(get % "title") (get page-entry "images"))) page-entries))
+        search-url (str "https://en.wikipedia.org/w/api.php?action=query&prop=images&format=json&titles=" (URLEncoder/encode search-query))
+        initial-result (-> search-url slurp json/read-str)]
+    (-> initial-result
+        (get-in ["query" "pages"])
+        vals
+        extract-titles)))
+
+(defn- extract-entries
+  [page-entries]
+  (-> (map (fn [page-entry]
+             {"title" (get page-entry "title")
+              "url" (-> page-entry
+                        (get "imageinfo")
+                        first
+                        (get "url"))})
+           page-entries)
+      vec
+      first))
+
+(defn- image-title->image-entry
+  [image-title]
+  (-> (str "https://en.wikipedia.org/w/api.php?action=query&prop=imageinfo&iiprop=url&format=json&titles=" (URLEncoder/encode image-title))
+      slurp
+      json/read-str
+      (get-in ["query" "pages"])
+      vals
+      extract-entries))
+
+(defn- image-search-handler
+  ""
+  [{:keys [query-params] :as request}]
+  (printlog request "query params:" query-params)
+  (try
+    (let [{:strs [search]} query-params
+          search-url (str "https://en.wikipedia.org/w/api.php?action=query&prop=images&format=json&titles=" search)
```

Comment:
Can we extract the parts of this url that are duplicated into a common place?

Reply:
`search-url` doesn't appear to be used in this function.

Reply:
Will remove.

Reply:
removed